project-lisa/source-code/client/src/api.ts

```diff
@@ -1,24 +1,261 @@
+*/
+export const raw = _git
+
+/**
+	* The http client for the git raw api. (Used for legacy migrations)
+	*
+	* Note: The http client is web-based. Node version 18 is
+	* required for the http client to work in node environments.
+*/
+export const http = _http
 
 export type Repository = {
-	fs: any
-	commit: () => unknown
-	push: () => unknown
-	pull: () => unknown
-	currentBranch: () => unknown
-	changeBranch: () => unknown
-	listRemotes: () => unknown
-	log: () => unknown
-	status: () => any
+	fs: NodeishFilesystem
+	commit: (args: { dir: string, author: any, message: string}) => Promise<Awaited<ReturnType<typeof raw.commit>> | undefined>
+	push: (args: { dir: string }) => Promise<Awaited<ReturnType<typeof raw.push>> | undefined>
+	pull: (args: { dir: string, author: any, fastForward: boolean, singleBranch: true }) => any
+	add: (args: { dir: string, filepath: string }) =>  Promise<Awaited<ReturnType<typeof raw.add>>>
+	listRemotes: () => Promise<Awaited<ReturnType<typeof raw.listRemotes>> | undefined>
+	log: (args: { dir: string, since: any}) =>  Promise<Awaited<ReturnType<typeof raw.log>>>
+	statusMatrix: (args: { dir:string, filter: any }) => Promise<Awaited<ReturnType<typeof raw.statusMatrix>>>
+	status: (args: { dir: string, filepath: string }) => Promise<Awaited<ReturnType<typeof raw.status>>>
+	mergeUpstream: (args: { branch: string }) => ReturnType<typeof github.request<"POST /repos/{owner}/{repo}/merge-upstream">>
+	isCollaborator: (args: { username: string }) => Promise<boolean>
+	getOrigin: () => Promise<string>
+	getCurrentBranch: () => Promise<string | undefined>
+	getMeta: () => Promise<{
+		name: string,
+		isPrivate: boolean,
+		isFork: boolean,
+		owner: { name?: string, email?: string },
+		parent:  {
+			url: string,
+			fullName: string
+		} | undefined
+	}>
+
+	// TODO: implement these before publishing api, but not used in badge or editor
+	// currentBranch: () => unknown
+	// changeBranch: () => unknown
+}
+
+export function load ({ url, fs, corsProxy }: { url: string, fs?: NodeishFilesystem, path?: string, corsProxy?: string, auth?: unknown }): Repository {
+	const rawFs = fs ? fs : createNodeishMemoryFs()
+
+	// parse url in the format of github.com/inlang/example and split it to host, owner and repo
+	const [host, owner, repoName] = [...url.split("/")]
+
+	// check if all 3 parts are present, if not, return an error
+	if (!host || !owner || !repoName) {
+		throw new Error(
+			`Invalid url format for '${url}' for cloning repository, please use the format of github.com/inlang/example.`,
+		)
+	}
+
+	const normalizedUrl = `https://${host}/${owner}/${repoName}`
+
+	let pending: Promise<void> | undefined = raw.clone({
+		fs: wrap(rawFs, 'clone'),
+		http,
+		dir: "/",
+		corsProxy,
+		url: normalizedUrl,
+		singleBranch: true,
+		depth: 1,
+		noTags: true,
+	}).finally(() => {
+		pending = undefined
+	}).catch((err: any) => {
+		console.error('error cloning the repository', err)
+	})
```

Comment:
Why don't you await the promise? 

If this is preparation for the lazy fetching, the correct solution is likely to ditch `raw.clone` entirely for a custom "clone" functionality that only fetches the minimum metadata required. Calls to `await fs.readFile()` fetch underlying contents of files lazily.